crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,408 @@ impl<'a> FileNoqaDirectives<'a> {
     }
 }
 
-/// An individual file-level exemption (e.g., `# ruff: noqa` or `# ruff: noqa: F401, F841`). Like
-/// [`FileNoqaDirectives`], but only for a single line, as opposed to an aggregated set of exemptions
-/// across a source file.
+/// Output of lexing a `noqa` directive.
 #[derive(Debug)]
-pub(crate) enum ParsedFileExemption<'a> {
-    /// The file-level exemption ignores all rules (e.g., `# ruff: noqa`).
-    All,
-    /// The file-level exemption ignores specific rules (e.g., `# ruff: noqa: F401, F841`).
-    Codes(Codes<'a>),
+pub(crate) struct NoqaLexerOutput<'a> {
+    warnings: Vec<LexicalWarning>,
+    directive: Directive<'a>,
 }
 
-impl<'a> ParsedFileExemption<'a> {
-    /// Return a [`ParsedFileExemption`] for a given `comment_range` in `source`.
-    fn try_extract(comment_range: TextRange, source: &'a str) -> Result<Option<Self>, ParseError> {
-        let line = &source[comment_range];
-        let offset = comment_range.start();
-        let init_line_len = line.text_len();
+/// Lexes in-line `noqa` comment, e.g. `# noqa: F401`
+fn lex_inline_noqa(
+    comment_range: TextRange,
+    source: &str,
+) -> Result<Option<NoqaLexerOutput<'_>>, LexicalError> {
+    let source = &source[..comment_range.end().to_usize()];
+    let offset = comment_range.start();
+    let lexer = NoqaLexer::starts_at(offset, source);
+    lexer.lex_inline_noqa()
```

Comment:
Nit: Both `lex_inline_noqa` and `lex_file_exemption` need to do the slicing and range selection here. Should we add a `NoqaLexer::range(full_source, range)` or `NoqaLexer::in_range(full_source, range)` method instead?